test/rgw: use restler fuzzing tool

Suyashd999 opened the pull request:
[RESTler](https://github.com/microsoft/restler-fuzzer) is a stateful REST API fuzzing tool developed by Microsoft Research. It is designed to automatically test REST APIs by generating sequences of requests that follow the API's logical usage patterns. RESTler analyzes the API specification to understand dependencies between requests and builds complex request sequences that can uncover deep bugs in the backend systems.

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

Suyashd999 committed:
added original s3 bucket OpenAPI specification file

Signed-off-by: Suyash Dongre <suyashd999@gmail.com>

yuvalif commented:
there are many changes in the `s3_bucket_.json` file that are unrelated to linking the commands.
this is making the review more difficult. are these chanmges necessary?

pritha-srivastava commented on src/test/rgw/restler/README.MD:
````diff
@@ -0,0 +1,75 @@
+  --no_ssl
+```
+
+Note: Ensure that engine_settings.json is configured correctly. You can refer to this example: [engine settings reference](https://github.com/microsoft/restler-fuzzer/pull/962)
````
I am unable to find the engine_settings.json file in PR: https://github.com/microsoft/restler-fuzzer/pull/962

Suyashd999 committed:
Remove x-amz-security-token header from OpenAPI spec

The x-amz-security-token header was causing authentication failures
in RGW when STS auth engine attempted to decrypt session tokens.
This header should only be present when using AWS STS (Security
Token Service) authentication, but was being included in requests
even when STS was not configured, leading to "Decryption failed"
errors.

Removed the header to prevent unnecessary STS authentication
attempts when rgw_s3_auth_use_sts is not enabled.

Signed-off-by: Suyash Dongre <suyashd999@gmail.com>

yuvalif commented on src/test/rgw/restler/openapi.yaml:
```diff
@@ -34,116 +34,8 @@ externalDocs:
-          - cn-northwest-1
-        default: cn-north-1
-    description: The Amazon S3 endpoint for China (Beijing) and China (Ningxia)
+  - url: 'http://localhost:8000'
```
this would work only with vstart. which is good (for now)
later on we would need to be able to easily replace that in teuthology

yuvalif commented:
i dont understand why you madethe change in b24287c0781201f21adc6b3a83af8f2af20f6db9 ?
removed many operations, changed descriptions etc.
lets keep all there is there even if it fails, we can still focus first at some basic operations that we want to cover

yuvalif commented:
> > i dont understand why you madethe change in [b24287c](https://github.com/ceph/ceph/commit/b24287c0781201f21adc6b3a83af8f2af20f6db9) ? removed many operations, changed descriptions etc. lets keep all there is there even if it fails, we can still focus first at some basic operations that we want to cover
> 
> Actually I did not change the description or anything it is a git diff problem.
> 
> The git isnt able to pin point changes that were made. It is the same issue with JSON .
> 
> We will have to make very small changes per commit for git to track.

also, no need to remove all of the operations that you are not linking yet.
we should not worry about these failures.

Suyashd999 committed:
Added examples for `/{Bucket}/` and `/{Bucket}/{key}/`

Signed-off-by: Suyash Dongre <suyashd999@gmail.com>

yuvalif commented on src/test/rgw/restler/openapi.yaml:
```diff
@@ -640,6 +640,22 @@ paths:
+            DeleteObject:
+              operationId: DeleteObject
+              parameters:
+                Bucket: $request.path.Bucket
```
you also need to link the DEL, GET and HEAD to the object key from `PutObject`

Suyashd999 committed:
Linked `/{Bucket}/` & `/{Bucket}/{Key}/` PUT, GET, HEAD, DEL operations

Signed-off-by: Suyash Dongre <suyashd999@gmail.com>

Suyashd999 committed:
Converted `/{Bucket}#policy` to Proxy compatible and linked `/{Bucket}?policy` PUT, GET, DEL operations

Signed-off-by: Suyash Dongre <suyashd999@gmail.com>

yuvalif commented on src/test/rgw/restler/README.MD:
there is no "restler_bin" directory here.
please update according to the install+build instructions above

Suyashd999 committed:
Add modular endpoint handler framework with policy fuzzing

Create extensible endpoint_handlers/ architecture for S3 API fuzzing:

- policy_handler/: S3 bucket policy fuzzing with templated-json-fuzzer
  * Normal and aggressive fuzzing modes with caching
  * Integration with s3_proxy_fuzz.py for transparent injection
  * Helper functions for policy generation and selection

- s3_proxy_fuzz.py: HTTP proxy with AWS SigV4 re-signing
  * Opt-in policy fuzzing via --enable-policy-fuzzing flag
  * Path normalization and configurable logging
  * Modular design for easy handler integration

Framework designed for easy expansion to support additional endpoints
(tagging, ACL, CORS, lifecycle, etc.)

Signed-off-by: Suyash Dongre <suyashd999@gmail.com>

adamemerson commented:
Still waiting for RESTler to merge support.
